# Практическая работа №3:  
## Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики

Выполнил студент гр. 9383 Ноздрин Василий. Вариант 13.

## Постановка задачи  

1. Решить вариант диофантова уравнения.  
1. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.  
1. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.  


## Выполнение работы

Решить уравнение $1883x-875y=-35$

In [1]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols

In [2]:
diophantine(1883*symbols('x') + 875*symbols('y') + 35)

{(125*t_0 + 230, -269*t_0 - 495)}

Это соответствует решению $$x=125c+230,\quad y = -269c-495,\quad c\in \mathbb{Z}$$

Русское слово из четырех букв закодировано алгоритмом RSA открытым ключом $(e=7, m=33)$. Шифрованное сообщение имеет вид $(7;4;12;13)$. Подберите закрытую часть ключа и прочитайте исходное слово. Буквами русского алфавита соответствуют числа в диапазоне от 2 до 32, исключая буквы 'ё' и 'ъ'.

In [3]:
from sympy.crypto.crypto import rsa_private_key, decipher_rsa

In [4]:
private_key = rsa_private_key(33, 7)
encoded_word = [7, 4, 12, 13]
alphabet = [chr(x) for x in range(ord('а'), ord('я')+1)]
alphabet.remove('ъ')
decoded_word = ''.join([
    alphabet[decipher_rsa(x, private_key)-2] for x in encoded_word
])
print(f"Приватный ключ: {private_key}\nСлово: {decoded_word}")

Приватный ключ: (33, 3)
Слово: люкс


Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением $f(x,y,z)=((z\vee y)(x\otimes y))\otimes(z(x\vee z))$.

In [5]:
from sympy.logic.boolalg import to_dnf, to_cnf

In [6]:
x, y, z = symbols('x y z')

f = ((z | y) & (x ^ y)) ^ (z & (x | z))

print(to_dnf(f,True), to_cnf(f,True), sep='\n')

(x & y & z) | (y & ~x & ~z) | (z & ~x & ~y)
(y | z) & (y | ~x) & (z | ~x) & (x | ~y | ~z)


Это соответствует выражениям $xyz\vee\overline{x}y\overline{z}\vee\overline{x}\overline{y}z, \quad (y\vee z)(\overline{x}\vee y)(\overline{x}z)(x\vee\overline{y}\vee\overline{z})$

## Вывод  

В ходе выполнения данной лабораторной работы была изучена библиотека символьных вычислений ***SymPy***. С помощью данной библиотеки были решены три задачи.